64bit Android builds fail due incorrect long double assumption

Hello,

while compiling Boost 1.91.0 for Android the following error occurs in Boost Math:

```
[2026-04-25T11:36:24.906Z] clang-linux.compile.c++ /home/opc/iwra973m9n7wtpfh/_output/Applications_Android_x86_64_cxx20/boost/bin.v2/libs/math/build/clang-linux-android/debug/x86_64/link-static/pch-off/target-os-android/threading-multi/visibility-global/atanh.o
[2026-04-25T11:36:24.906Z] In file included from /home/opc/iwra973m9n7wtpfh/libs/math/build/../src/tr1/atanh.cpp:11:
[2026-04-25T11:36:24.906Z] In file included from ./boost/math/special_functions/atanh.hpp:22:
[2026-04-25T11:36:24.906Z] In file included from ./boost/math/special_functions/log1p.hpp:34:
[2026-04-25T11:36:24.906Z] In file included from ./boost/math/special_functions/fpclassify.hpp:21:
[2026-04-25T11:36:24.906Z] ./boost/math/special_functions/detail/fp_traits.hpp:314:1: error: static_assert failed due to requirement '113 == 64' "Oops, assumption that long double is an 80-bit quantity is incorrect!!"
[2026-04-25T11:36:24.906Z] static_assert(LDBL_MANT_DIG == 64, "Oops, assumption that long double is an 80-bit quantity is incorrect!!");
[2026-04-25T11:36:24.906Z] ^             ~~~~~~~~~~~~~~~~~~~
[2026-04-25T11:36:24.906Z] 1 error generated.
```

Using Android NDK 25 I dumped the value of `LDBL_MANT_DIG` for all architectures:
```
$ $ANDROID_NDK/toolchains/llvm/prebuilt/linux-x86_64/bin/x86_64-linux-android21-clang -dM -E - < /dev/null |grep LDBL_MANT_DIG
#define __LDBL_MANT_DIG__ 113
$ $ANDROID_NDK/toolchains/llvm/prebuilt/linux-x86_64/bin/x86_64-linux-android33-clang -dM -E - < /dev/null |grep LDBL_MANT_DIG
#define __LDBL_MANT_DIG__ 113

$ $ANDROID_NDK/toolchains/llvm/prebuilt/linux-x86_64/bin/i686-linux-android19-clang -dM -E - < /dev/null |grep LDBL_MANT_DIG
#define __LDBL_MANT_DIG__ 53
$ $ANDROID_NDK/toolchains/llvm/prebuilt/linux-x86_64/bin/i686-linux-android33-clang -dM -E - < /dev/null |grep LDBL_MANT_DIG
#define __LDBL_MANT_DIG__ 53

$ $ANDROID_NDK/toolchains/llvm/prebuilt/linux-x86_64/bin/aarch64-linux-android21-clang -dM -E - < /dev/null |grep LDBL_MANT_DIG
#define __LDBL_MANT_DIG__ 113
$ $ANDROID_NDK/toolchains/llvm/prebuilt/linux-x86_64/bin/aarch64-linux-android33-clang -dM -E - < /dev/null |grep LDBL_MANT_DIG
#define __LDBL_MANT_DIG__ 113

$ $ANDROID_NDK/toolchains/llvm/prebuilt/linux-x86_64/bin/armv7a-linux-androideabi19-clang -dM -E - < /dev/null |grep LDBL_MANT_DIG
#define __LDBL_MANT_DIG__ 53
$ $ANDROID_NDK/toolchains/llvm/prebuilt/linux-x86_64/bin/armv7a-linux-androideabi33-clang -dM -E - < /dev/null |grep LDBL_MANT_DIG
#define __LDBL_MANT_DIG__ 53
```

For each architecture I used the oldest and latest ABI version. As you can see the 64 bit Intel and ARM architectures use `113` instead of the expected `64`. If you need something to differentiate the Android platform, the `__ANDROID__` preprocessor symbol is set to 1 within the toolchain defaults.

Thanks,
Gregor

--- a/.drone/drone.sh
+++ b/.drone/drone.sh
@@ -51,8 +51,8 @@ echo '==================================> BEFORE_SCRIPT'
 echo '==================================> SCRIPT'
 
 echo "using $TOOLSET : : $COMPILER : <cxxflags>-std=$CXXSTD $OPTIONS ;" > ~/user-config.jam
-(cd libs/config/test && ../../../b2 config_info_travis_install toolset=$TOOLSET && ./config_info_travis)
-(cd libs/math/test && ../../../b2 toolset=$TOOLSET $TEST_SUITE define=$DEFINE)
+(cd libs/config/test && ../../../b2 print_config_info print_math_info toolset=$TOOLSET)
+(cd libs/math/test && ../../../b2 -j3 toolset=$TOOLSET $TEST_SUITE)
 
 echo '==================================> AFTER_SUCCESS'
 

--- a/include/boost/math/special_functions/detail/fp_traits.hpp
+++ b/include/boost/math/special_functions/detail/fp_traits.hpp
@@ -270,13 +270,7 @@ template<> struct fp_traits_non_native<double, double_precision>
 
 // long double (64 bits) -------------------------------------------------------
 
-#if defined(BOOST_NO_INT64_T) || defined(BOOST_NO_INCLASS_MEMBER_INITIALIZATION)\
-   || defined(BOOST_BORLANDC) || defined(__CODEGEAR__) || (defined(__APPLE__) && defined(__aarch64__)) || defined(_MSC_VER)\
-    || (defined(__GNUC__) && defined(__aarch64__) && defined(_WIN32))\
-     || (defined(__GNUC__) && (defined(__arm__) || defined(__thumb__)))\
-      || defined(__SYCL_DEVICE_ONLY__) || (LDBL_MANT_DIG == 53)
-
-static_assert(LDBL_MANT_DIG == 53, "Oops, assumption that long double is a 64-bit quantity is incorrect!!");
+#if LDBL_MANT_DIG == 53 || defined(__SYCL_DEVICE_ONLY__)
 
 template<> struct fp_traits_non_native<long double, double_precision>
 {
@@ -305,16 +299,9 @@ private:
 
 // long double (>64 bits), x86 and x64 -----------------------------------------
 
-// Andorid is a known exception because on x64 it reports LDBL_MANT_DIG == 113
-// See: https://github.com/boostorg/math/issues/1389
-
-#elif (defined(__i386) || defined(__i386__) || defined(_M_IX86) \
-    || defined(__amd64) || defined(__amd64__)  || defined(_M_AMD64) \
-    || defined(__x86_64) || defined(__x86_64__) || defined(_M_X64)) && !defined(__ANDROID__)
+#elif LDBL_MANT_DIG == 64
 
 // Intel extended double precision format (80 bits)
-
-static_assert(LDBL_MANT_DIG == 64, "Oops, assumption that long double is an 80-bit quantity is incorrect!!");
 
 template<>
 struct fp_traits_non_native<long double, extended_double_precision>
@@ -442,8 +429,6 @@ struct fp_traits_non_native<long double, extended_double_precision>
 #elif (LDBL_MANT_DIG == 113)
 
 // IEEE extended double precision format with 15 exponent bits (128 bits)
-
-static_assert(LDBL_MANT_DIG == 113, "Oops, assumption that long double is a 128-bit quantity is incorrect!!");
 
 template<>
 struct fp_traits_non_native<long double, extended_double_precision>

--- a/include/boost/math/special_functions/detail/fp_traits.hpp
+++ b/include/boost/math/special_functions/detail/fp_traits.hpp
@@ -305,9 +305,12 @@ private:
 
 // long double (>64 bits), x86 and x64 -----------------------------------------
 
-#elif defined(__i386) || defined(__i386__) || defined(_M_IX86) \
+// Andorid is a known exception because on x64 it reports LDBL_MANT_DIG == 113
+// See: https://github.com/boostorg/math/issues/1389
+
+#elif (defined(__i386) || defined(__i386__) || defined(_M_IX86) \
     || defined(__amd64) || defined(__amd64__)  || defined(_M_AMD64) \
-    || defined(__x86_64) || defined(__x86_64__) || defined(_M_X64)
+    || defined(__x86_64) || defined(__x86_64__) || defined(_M_X64)) && !defined(__ANDROID__)
 
 // Intel extended double precision format (80 bits)
 

